client/tinychain/ml/dnn.py

```diff
@@ -12,32 +12,118 @@ def layer(weights, bias, activation):
+            
+            '''
+            gradients function :it calculates delta values in Weights , Bias and in previous layer.
+            A_prev : Previous Input.
+            Z : Current Output before Activation.
+            dA: detla_Error.
+            dZ : delta Error in the Current Layer.
+            dA_Prev: delta Error in Previous Layer.
+            d_weights: delta Weights.
+            d_bias : delta Bias.
+            activation.backward(dA,Z) : performs derivative version of your activation function
+                                          to apply backward propagation.
+            
+            '''
             dZ = activation.backward(dA, Z).copy()
-            dA_prev = einsum("kj,ij->ki", [dZ, self[0]])
+            dA_prev = einsum("kj,ij->ki", [dZ, self[0]])  # Recommend self.weights
             d_weights = einsum("kj,ki->ij", [dZ, A_prev])
             d_bias = dZ.sum(0)
             return dA_prev, d_weights, d_bias
 
         def train_eval(self, inputs):
+            '''
+            train_eval function you can use it on any layer to get its
+            output after dot product and after applying activation function on that output
+            
+            Z : Output of dot product between weight and inputs .
+            A : returning the output after applying activation function on it.
+            '''
             Z = einsum("ij,ki->kj", [self.weights, inputs])
             A = activation.forward(Z) + self.bias
             return A, Z
 
         def update(self, d_weights, d_bias):
+            '''
+            update function is used to simply update weights and bias and return new weights and bias
+            New_Weight = Old Weight - Delta Weight
+            New_Bias = Old Bias - Delta Bias
+            '''
             return self.weights.write(self.weights - d_weights), self.bias.write(self.bias - d_bias)
+        
+        
+        
+        def adam(self, A_prev, dA, Z ,decay_rate_1 : tc.Number, decay_rate_2 : tc.Number, 
+                 epsilon :tc.Number , l_m:tc.Number,l_v:tc.Number,t:tc.Number]) :  
+                             
+            dZ = activation.backward(dA, Z).copy()          
+            # Gradients for each layer
+            g = einsum("kj,ki->ij", [dZ, A_prev])  
+            
+            t=t+1
+            
+            # Computing 1st and 2nd moment for each layer
+            l_m = l_m * decay_rate_1 + (1- decay_rate_1) * g            
+            l_v = l_v * decay_rate_2 + (1- decay_rate_2) * (g ** 2)            
+            l_m_corrected = l_m / (1-(decay_rate_1 ** t))
+            l_v_corrected = l_v / (1-(decay_rate_2 ** t))
+            
+            # Update Weights
+            d_weights = l_m_corrected / ((l_v_corrected)**(0.5) + epsilon)                       
+            d_bias = dZ.sum(0)
+            
+            return l_m,l_v,t,self.weights.write(self.weights - d_weights), self.bias.write(self.bias - d_bias)
```

Comment:
I think a map would be more clear here, like
`return tc.After(updates, new_state)`
Where new_state is a map with the state of the optimizer